### PR TITLE
make oidc authenticator (more?) audience aware

### PR DESCRIPTION
**Pull request by mikedanese**
Part of #69893

/kind feature
/sig auth

@kubernetes/sig-auth-pr-reviews 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ type Options struct {
 	// The plugin supports the "authorized party" OpenID Connect claim, which allows
 	// specialized providers to issue tokens to a client for a different client.
 	// See: https://openid.net/specs/openid-connect-core-1_0.html#IDToken
-	ClientID string
+	ClientIDs authenticator.Audiences
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Ok I won't sneak in mutliple audiences yet...

File: pkg/kubeapiserver/authenticator/config.go [hunk 2]
```diff
@@ -238,7 +238,7 @@ func newAuthenticatorFromTokenFile(tokenAuthFile string) (authenticator.Token, e
 }
 
 // newAuthenticatorFromOIDCIssuerURL returns an authenticator.Token or an error.
-func newAuthenticatorFromOIDCIssuerURL(issuerURL, clientID, caFile, usernameClaim, usernamePrefix, groupsClaim, groupsPrefix string, signingAlgs []string, requiredClaims map[string]string) (authenticator.Token, error) {
+func newAuthenticatorFromOIDCIssuerURL(issuerURL, clientID, caFile, usernameClaim, usernamePrefix, groupsClaim, groupsPrefix string, signingAlgs []string, requiredClaims map[string]string, apiAudiences authenticator.Audiences) (authenticator.Token, error) {
```
Comment (awly) [inline pkg/kubeapiserver/authenticator/config.go]: Can you pass in oidc.Options instead?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 3]
```diff
@@ -78,6 +78,12 @@ type Options struct {
 	// See: https://openid.net/specs/openid-connect-core-1_0.html#IDToken
 	ClientID string
 
+	// APIAudiences are the implicit audiences that the API server identitifes
```
Comment (awly) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Why "implicit"?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 4]
```diff
@@ -532,6 +542,9 @@ func (r *claimResolver) resolve(endpoint endpoint, allClaims claims) error {
 }
 
 func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+	if reqAuds, ok := authenticator.AudiencesFrom(ctx); ok && len(reqAuds.Intersect(a.clientIDs)) == 0 && len(reqAuds.Intersect(a.apiAudiences)) == 0 {
```
Comment (awly) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Move the intersect checks into an inner if, it's hard to parse

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go [hunk 5]
```diff
@@ -1388,6 +1396,149 @@ func TestToken(t *testing.T) {
 				Name: "thomas.jefferson@gmail.com",
 			},
 		},
+		// tests that muck with the audiences
```
Comment (awly) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go]: remove the comment, test names are descriptive enough

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 3 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Dropped. We've used implicit to refer to the server audience identifiers when to target audience is provided as part of the authentication request. It is unnecessary and confusing to include that here.

File: pkg/kubeapiserver/authenticator/config.go (hunk 2 above)
Comment (mikedanese) [reply pkg/kubeapiserver/authenticator/config.go]: Done.

**Comment by mikedanese** [comment]
@awly Fixed. PTAL.
